src/net/discovery/mod.rs

```diff
@@ -0,0 +1,168 @@
+pub mod conf;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    ParseError(#[from] config::ConfigError),
+    #[error(transparent)]
+    IOError(#[from] std::io::Error),
+}
+
+pub mod peer {
+    use axum::http::{
+        uri::{Authority, Scheme},
+        Uri,
+    };
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+    use std::fmt::Formatter;
+    use std::ops::{Deref, DerefMut};
+
+    /// Describes just the origin of a url, i.e.: "http\[s\]://\[authority\]", minus the path and
```

Comment:
I look at this again and maybe we should just assume that helper addr is a url? that will remove all of this complexity of serializing from here

Reply:
what type are you proposing we use? `Uri` itself is not serializable, so we would still need to handle that. we could do string, and maybe just check that it's valid after init?

Reply:
yea I like using str here. less code is better. Parsing a url and validating it should be 2 lines or code

Reply:
my only concern at that point is that all implementers of `PeerDiscovery` will be responsible for validating the string. so you'd have 2 lines of code per implementer, as opposed to having this code exist in one place